### PR TITLE
Avoid `LogicalPlan::clone()` in `LogicalPlan::map_children` when possible

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

This is the first step towards making DataFusion planning (much) faster -- https://github.com/apache/arrow-datafusion/issues/9637 (based on ideas from https://github.com/apache/arrow-datafusion/pull/9708 and https://github.com/apache/arrow-datafusion/pull/9768). 🙏 @jayzhan211

## Rationale for this change

Profiling (see https://github.com/apache/arrow-datafusion/issues/9637) shows that slowest part of planning is copying `LogicalPlan` and `Expr` in the `Optimizer` 

Thus avoiding this copying as much as possible is key.

The TreeNode API is fast becoming the standard way to rewrite `LogicalPlan` in DataFusion so it is worth investing time making them faster (so everything implemented in terms of them gets faster). When I rewrote the `Optimizer` to use this API (in https://github.com/apache/arrow-datafusion/pull/9948 -- it gets 10% faster)

## What changes are included in this PR?
1. Update `LogicalPlan::map_children` to avoid cloning inputs when possible (by rewriting `Arc` as much as possible)

## Are these changes tested?

Functionally: By existing CI

Performance benchmarks: 1-2% faster in planning benchmarks (this only the beginning! Subsequent PRs are even better)

<details><summary>Details</summary>
<p>

```
++ critcmp main map_in_place2
group                                         main                                   map_in_place2
-----                                         ----                                   -------------
logical_aggregate_with_join                   1.00  1182.7±63.77µs        ? ?/sec    1.00  1183.5±12.43µs        ? ?/sec
logical_plan_tpcds_all                        1.00    154.4±0.63ms        ? ?/sec    1.00    154.4±0.65ms        ? ?/sec
logical_plan_tpch_all                         1.00     16.6±0.15ms        ? ?/sec    1.00     16.6±0.14ms        ? ?/sec
logical_select_all_from_1000                  1.00     19.2±0.13ms        ? ?/sec    1.00     19.3±0.09ms        ? ?/sec
logical_select_one_from_700                   1.00    774.1±7.20µs        ? ?/sec    1.00    774.9±6.98µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.00    723.6±6.95µs        ? ?/sec    1.01   730.1±12.65µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.00    711.2±9.34µs        ? ?/sec    1.01   718.5±26.52µs        ? ?/sec
physical_plan_tpcds_all                       1.02   1831.7±3.37ms        ? ?/sec    1.00   1799.9±9.30ms        ? ?/sec
physical_plan_tpch_all                        1.01    119.1±0.61ms        ? ?/sec    1.00    117.5±0.56ms        ? ?/sec
physical_plan_tpch_q1                         1.02      7.3±0.10ms        ? ?/sec    1.00      7.1±0.05ms        ? ?/sec
physical_plan_tpch_q10                        1.02      5.5±0.04ms        ? ?/sec    1.00      5.4±0.02ms        ? ?/sec
physical_plan_tpch_q11                        1.02      4.8±0.03ms        ? ?/sec    1.00      4.7±0.02ms        ? ?/sec
physical_plan_tpch_q12                        1.02      3.9±0.04ms        ? ?/sec    1.00      3.8±0.02ms        ? ?/sec
physical_plan_tpch_q13                        1.01      2.6±0.01ms        ? ?/sec    1.00      2.6±0.01ms        ? ?/sec
physical_plan_tpch_q14                        1.02      3.3±0.01ms        ? ?/sec    1.00      3.3±0.01ms        ? ?/sec
physical_plan_tpch_q16                        1.01      4.8±0.02ms        ? ?/sec    1.00      4.8±0.02ms        ? ?/sec
physical_plan_tpch_q17                        1.01      4.6±0.02ms        ? ?/sec    1.00      4.6±0.02ms        ? ?/sec
physical_plan_tpch_q18                        1.02      5.0±0.04ms        ? ?/sec    1.00      4.9±0.02ms        ? ?/sec
physical_plan_tpch_q19                        1.00      9.4±0.09ms        ? ?/sec    1.00      9.3±0.06ms        ? ?/sec
physical_plan_tpch_q2                         1.01     10.5±0.08ms        ? ?/sec    1.00     10.4±0.07ms        ? ?/sec
physical_plan_tpch_q20                        1.02      6.1±0.04ms        ? ?/sec    1.00      6.0±0.03ms        ? ?/sec
physical_plan_tpch_q21                        1.02      8.3±0.04ms        ? ?/sec    1.00      8.1±0.06ms        ? ?/sec
physical_plan_tpch_q22                        1.03      4.4±0.01ms        ? ?/sec    1.00      4.3±0.02ms        ? ?/sec
physical_plan_tpch_q3                         1.01      3.9±0.01ms        ? ?/sec    1.00      3.8±0.02ms        ? ?/sec
physical_plan_tpch_q4                         1.03      2.9±0.03ms        ? ?/sec    1.00      2.8±0.02ms        ? ?/sec
physical_plan_tpch_q5                         1.01      5.6±0.02ms        ? ?/sec    1.00      5.5±0.04ms        ? ?/sec
physical_plan_tpch_q6                         1.02  1975.5±13.85µs        ? ?/sec    1.00  1944.6±11.61µs        ? ?/sec
physical_plan_tpch_q7                         1.01      7.5±0.03ms        ? ?/sec    1.00      7.4±0.03ms        ? ?/sec
physical_plan_tpch_q8                         1.01      9.5±0.06ms        ? ?/sec    1.00      9.4±0.04ms        ? ?/sec
physical_plan_tpch_q9                         1.01      7.2±0.04ms        ? ?/sec    1.00      7.1±0.03ms        ? ?/sec
physical_select_all_from_1000                 1.00    128.0±0.45ms        ? ?/sec    1.00    128.0±0.45ms        ? ?/sec
physical_select_one_from_700                  1.00      4.0±0.02ms        ? ?/sec    1.00      4.0±0.02ms        ? ?/sec```

</p>
</details> 

## Are there any user-facing changes?
(slightly) faster Optimizer performance 



## Notes

Huge thanks to @peter-toth  for sorting out and driving the `TreeNode` API in the first place, most recently https://github.com/apache/arrow-datafusion/pull/9913

Note this is different than previous designs (e.g. https://github.com/apache/arrow-datafusion/pull/9946) where a default `Arc` was left in place. Thanks to the comments from @peter-toth  and @jayzhan211  I think this API is now much cleaner than earlier versions.

### CODE REVIEW

File: datafusion/expr/src/logical_plan/tree_node.rs [hunk 1]
```diff
@@ -32,23 +39,368 @@ impl TreeNode for LogicalPlan {
...
+    ///
+    /// # Notes
+    ///
+    /// Inputs include ONLY direct children, not embedded `LogicalPlan`s for
+    /// subqueries, for example such as are in [`Expr::Exists`].
+    fn map_children<F>(self, mut f: F) -> Result<Transformed<Self>>
+    where
+        F: FnMut(Self) -> Result<Transformed<Self>>,
+    {
+        Ok(match self {
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/tree_node.rs:52]: This follows the (really very cool) pattern @peter-toth came up with in https://github.com/apache/arrow-datafusion/pull/9913

File: datafusion/expr/src/logical_plan/tree_node.rs [hunk 2]
```diff
@@ -32,23 +39,362 @@ impl TreeNode for LogicalPlan {
...
+            LogicalPlan::TableScan { .. }
+            | LogicalPlan::Statement { .. }
+            | LogicalPlan::EmptyRelation { .. }
+            | LogicalPlan::Values { .. }
+            | LogicalPlan::DescribeTable(_) => Transformed::no(self),
+        })
     }
 }
+
+/// Converts a `Arc<LogicalPlan>` without copying, if possible. Copies the plan
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/tree_node.rs:338]: Here is the code that avoids copying for `Arc<LogicalPlan>` when possible (my performance results show it is possible most of the time)

**Review by peter-toth — APPROVED** [review]
LGTM, thanks for the idea and the implementation @alamb and @jayzhan211!

**Comment by alamb** [comment]
> LGTM, thanks for the idea and the implementation @alamb and @jayzhan211!

Thanks @peter-toth  -- I agree I am quite happy with this formulation compared to the initial versions. It is pretty slick and a great example of collaboration I think. 

I am now very excited to use it to stop the copying in the Optimizer -- I like refactoring as much as the next person, but it really helps to have an user visible goal (faster planning)

**Review by jayzhan211 — APPROVED** [review]
I agree that we can avoid unwrap and wrap with Arc if the transform state is `TreeNodeRecursion::Stop`

File: datafusion/expr/src/logical_plan/tree_node.rs [hunk 3]
```diff
@@ -32,23 +39,364 @@ impl TreeNode for LogicalPlan {
...
+fn rewrite_arcs<F>(
+    input_plans: Vec<Arc<LogicalPlan>>,
+    mut f: F,
+) -> Result<Transformed<Vec<Arc<LogicalPlan>>>>
+where
+    F: FnMut(LogicalPlan) -> Result<Transformed<LogicalPlan>>,
+{
+    Ok(input_plans
+        .into_iter()
+        .map(unwrap_arc)
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/expr/src/logical_plan/tree_node.rs]: This is an excellent suggestion -- the code is both less verbose, and and it saves having to make a second `Vec`

I made this change in b29ebd26e

**Comment by alamb** [comment]
Thanks @peter-toth  and @jayzhan211

**Pull request merged**